cin_validator/rules/cin2022_23/rule_8590.py

```diff
@@ -0,0 +1,165 @@
+from typing import Mapping
+
+import pandas as pd
+
+from cin_validator.rule_engine import CINTable, RuleContext, rule_definition
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+ChildIdentifiers = CINTable.ChildIdentifiers
+CINdetails = CINTable.CINdetails
+
+LAchildID = ChildIdentifiers.LAchildID
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here, in place of 2885
+    code=8590,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.ChildIdentifiers,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="Child does not have a recorded CIN episode.",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[LAchildID],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cid = data_container[ChildIdentifiers].copy()
+    df_cin = data_container[CINdetails].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cid.index.name = "ROW_ID"
+    df_cin.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cid.reset_index(inplace=True)
+    df_cin.reset_index(inplace=True)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # Each child must have at least one <CINdetails> group
+
+    df_merge = df_cid.merge(
+        df_cin[LAchildID], on=[LAchildID], how="left", indicator=True
+    )
+    # get only ChildIdentifiers that don't have a matching CINdetails.
+    condition = df_merge["_merge"] == "left_only"
```

Comment:
I can't say I've ever seen this, it's very interesting.

Reply:
I think merge indicator is used quite a lot in the SSDA903 tool rules so you'll see a lot more of it when you look over there! :)